File: Algorithms/p239_Sliding_Window_Maximum/p239_Sliding_Window_Maximum_2.py
class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        res = []
        heap = [None]*k # heap[i]: nums index
        ptr = [None]*len(nums) # ptr[i]: heap index of nums[i]
        for i in range(0,len(nums)):
            if i>=k:
                # remove nums[i-k] from heap
                heap[ptr[i-k]] = heap[k-1]
                ptr[heap[k-1]] = ptr[i-k]
                self.heapify(nums,heap,k,ptr,ptr[i-k])
            # add nums[i] to heap
            j = min(i,k-1)
            heap[j] = i
            ptr[i] = j
            self.heapify(nums,heap,j+1,ptr,j)
            if i>=k-1:
                res.append(nums[heap[0]])
        return res

    def heapify(self, nums, heap, size, ptr, i):
        while i>0:
            p = (i-1)//2
            if nums[heap[i]]>nums[heap[p]]:
                heap[i],heap[p] = heap[p],heap[i]
                ptr[heap[i]],ptr[heap[p]] = ptr[heap[p]],ptr[heap[i]]
                i = p
            else:
                break
        while i<size:
            c1 = i*2+1
            c2 = i*2+2
            if c1<size and nums[heap[i]]<nums[heap[c1]] and (c2>=size or nums[heap[c1]]>=nums[heap[c2]]):
                heap[i],heap[c1] = heap[c1],heap[i]
                ptr[heap[i]],ptr[heap[c1]] = ptr[heap[c1]],ptr[heap[i]]
                i = c1
            elif c2<size and nums[heap[i]]<nums[heap[c2]] and nums[heap[c2]]>=nums[heap[c1]]:
                heap[i],heap[c2] = heap[c2],heap[i]
                ptr[heap[i]],ptr[heap[c2]] = ptr[heap[c2]],ptr[heap[i]]
                i = c2
            else:
                break

File: Algorithms/p239_Sliding_Window_Maximum/p239_Sliding_Window_Maximum_2.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> res;
        vector<int> heap(k); // heap[i]: nums index
        vector<int> ptr(nums.size()); // ptr[i]: heap index of nums[i]
        for(int i=0;i<nums.size();++i){
            if(i>=k){
                // remove nums[i-k] from heap
                heap[ptr[i-k]] = heap[k-1];
                ptr[heap[k-1]] = ptr[i-k];
                heapify(nums,heap,k,ptr,ptr[i-k]);
            }
            // add nums[i] to heap
            int j = min(i,k-1);
            heap[j] = i;
            ptr[i] = j;
            heapify(nums,heap,j+1,ptr,j);
            if(i>=k-1) res.push_back(nums[heap[0]]);
        }
        return res;
    }
    void heapify(vector<int>& nums, vector<int>& heap, int size, vector<int>& ptr, int i) {
        while(i>0){
            int p = (i-1)/2;
            if(nums[heap[i]]>nums[heap[p]]){
                swap(heap[i],heap[p]);
                swap(ptr[heap[i]],ptr[heap[p]]);
                i = p;
            }else{
                break;
            }
        }
        while(i<size){
            int c1 = i*2+1;
            int c2 = i*2+2;
            if(c1<size&&nums[heap[i]]<nums[heap[c1]]&&(c2>=size||nums[heap[c1]]>=nums[heap[c2]])){
                swap(heap[i],heap[c1]);
                swap(ptr[heap[i]],ptr[heap[c1]]);
                i = c1;
            }else if(c2<size&&nums[heap[i]]<nums[heap[c2]]&&nums[heap[c2]]>=nums[heap[c1]]){
                swap(heap[i],heap[c2]);
                swap(ptr[heap[i]],ptr[heap[c2]]);
                i = c2;
            }else{
                break;
            }
        }
    }
};
